lib/ringsc.gi

```diff
@@ -403,14 +396,53 @@ InstallGlobalFunction( RingByStructureConstants, function( arg )
         NewType( Fam, IsSCRingObj and IsDenseCoeffVectorRep );
 
     SetCoefficientsFamily( Fam, ElementsFamily( FamilyObj( Integers ) ) );
-    # temporary
-    SetIsUFDFamily(Fam,false);
 
     # Make the generators and the ring.
     SetZero( Fam, ObjByExtRep( Fam, List( [ 1 .. n ], x -> 0 ) ) );
     gens:= Immutable( List( IdentityMat( n, Integers ),
                             x -> ObjByExtRep( Fam, x ) ) );
-    A:= RingByGenerators( gens );
+
+    # Hope to find a unit (and thus an identity)
+    one:=fail;
+    # Try generators first
+    i:=1;
+    while one=fail and i<=Length(gens) do
+      A:=[];
+      new:=gens[i];
+      repeat
+        Add(A,new);
+        new:=new*gens[i];
+      until new in A;
+      if new=gens[i] then
+        new:=A[Length(A)];
+        if ForAll(gens,x->x*new=x and new*x=x) then
+          one:=new;
+        fi;
+      fi;
+      i:=i+1;
+    od;
+
+    if one=fail then
+      # if not too big go through elements
+      if Product(moduli)<=10^5 then
+        A:=RingByGenerators(gens);
+        A:=Enumerator(A);
+        one:=First(A,x->ForAll(gens,y->y*x=y and x*y=y));
+      else
+        Info(InfoWarning,1,"currently cannot test for ring-with-one property");
+      fi;
+    fi;
+
+    # temporary
+    SetIsUFDFamily(Fam,false);
+
+    if one<>fail then
+      SetOne(Fam,one);
+      A:= RingWithOneByGenerators( gens );
+      SetOne(A,one);
+    else
+      A:= RingByGenerators( gens );
```

Comment:
So now we sometimes get a ring-with-one, and sometimes a ring, depending on a heuristic the user can not control; potentially changing the outcome of existing computations and tests?

This may work well in your concrete situation, but might it not cause problems in other situations that then are hard to compensate? I am worried we'll again see failing package tests and then someone has to waste time dealing with *that*.

How about instead adding an (optional?) argument that indicates whether the constructed ring should be a ring-with-one? Then the behaviour could be:
- if the argument is not given, or is set to `false, we construct a ring as before
- if the argument is given, and set to `true`, we use your heuristic, and produce an error if it fails
- optional: if the argument is set to `"try"`, we do what your PR does right now (i.e. try to find a one, but if that fails, just go on) -- while I am surprised that this is useful, it clearly is to you, so we can of course support it
- if the argument is set a "description of a ring element" (probably a vector relative to the base of the SC ring?), we assume that element is the identity element, and use that (possibly after validating it is an identity).


Wonder if @ChrisJefferson @ThomasBreuer have any thoughts about this?

Reply:
If I understand the proposal right then the function `RingByStructureConstants` shall decide whether its return value is a ring or ring-with-one, depending on some heuristics. An analogon in the world of multiplicative structures would be `SemigroupByMultiplicationTable` returning sometimes a monoid and sometimes just a semigroup.

Do we really want this?
This looks to me like asking for trouble, from a conceptual point of view.

Why can't we just introduce `RingWithOneByStructureConstants`, which takes the description of the identity element as a separate argument? In situations where one knows that the result should be regarded as a ring-with-one, one can call this function, and otherwise one first creates the ring and then asks for `AsRingWithOne` --and if that is expected to be too expensive then one could introduce `AsRingWithOneCheap` or so, which just checks some heuristics and gives up if this fails. (Note that the underlying family setup and element arithmetic of the original ring can be kept.)

There are similar situations in the GAP library which could be improved in the same direction:
Currently we do not have `AlgebrawithOneByStructureConstants` and `DivisionRingByStructureConstants`.
(There is an old comment in the function `QuaternionAlgebra` that `AlgebrawithOneByStructureConstants` would be useful, and I have code for that somewhere, but I admit that I never tried to add this to the GAP library.)

Reply:
> An analogon in the world of multiplicative structures would be SemigroupByMultiplicationTable returning sometimes a monoid and sometimes just a semigroup.

I think the answer is "no, we don't want this", although it was probably me that introduced this behaviour in the first place (sorry!). The reason for doing so was that most users seem to expect `IsMonoid` to return `true` if the semigroup they've defined is mathematically a monoid, not whether it belongs to the category of monoids (in the gap sense). However, in retrospect, as has been said elsewhere in this thread, using a function like `Semigroup` or `SemigroupByMultiplicationTable` in other code is harder than it should be because some times it returns a semigroup and sometimes a monoid, and so the reasoning that goes in the calling code is more complicated.

Reply:
I never understood why `WithOne` is not a property, but a separate category (but I also never understood why one would not adjoin an identity to semigroups if needed, so we only have monoids, so I might be missing a concept).

The main issue for allowing a ring with one if possible at all  is that some functionality absolutely requires a ring in the WithOne category. The existing test is bad (I could not come up with a better one, though there surely must be), and the limit was just to avoid a calculation not terminating. We could put this instead behind an option (though I would prefer the option to skip the check, rather than trigger it).

Reply:
Concerning the difference between `Something` and `SomethingWithOne`, the general idea is that different concepts of generation imply different categories in the sense of GAP's `IsCategory`.

Here is a situation where one runs into difficulties if one does not keep the two concepts separate.
Consider the idempotent `m:= [ [ 1, 0 ], [ 0, 0 ] ]`, a 2x2 matrix.
The algebra generated by `m` is 1-dimensional, it has a multiplicative identity, but it does not contain the identity matrix `m^0 = One( m )`.
The algebra-with-one generated by `m` is 2-dimensional, it contains also `m^0`.

Formally extending a semigroup or an algebra by an identity is no problem in theory, but in a concrete implementation one will in general create new objects and an embedding morphism.

What is the problem with explicitly calling `RingWithOneByStructureConstants` if one needs a ring-with-one?
- One knows that there is an identity but one does not know which element it is, it is expensive to compute it, and in fact one does not really need this identity in the computations?
- Or one has a large number of rings, many of them contain an identity, and if one knows this then certain computations are much cheaper? (For example, the generic `RadicalOfAlgebra` method distinguishes the two situations.)

Reply:
> Here is a situation where one runs into difficulties if one does not keep the two concepts separate.

Sure. `RingByGenerators` or `RingWithOneByGenerators` are different things. My puzzling goes towards making this part of the category of the created object, rather than a property. An object created as `RingByGenerators` can never be a `RingWithOne`, even if a `One` is found.

> What is the problem with explicitly calling RingWithOneByStructureConstants if one needs a ring-with-one?

The problem is basically that when starting to create rings through structure constants, one does not realize that one needs to make the ring explicitly a ring-with-one, so that basic functionality for matrices or polynomials works.

Reply:
Relevant question on MSE:

https://math.stackexchange.com/questions/4756076/viewing-monoid-rings-as-rings-with-identity-in-gap

Reply:
Concerning the remark

> My puzzling goes towards making this part of the category of the created object, rather than a property.

The relevant part of the GAP documentation is [the section on Operational Structure of Domains](https://docs.gap-system.org/doc/ref/chap31_mj.html#X859C7AB97B34F55F), in particular the following sentence.

> One important difference between the operational structure and the properties of a domain is that the operational structure is fixed when the domain is constructed, whereas properties can be discovered later.

For a domain D, having an identity (meaning that each element has an identity, it is the same identity for all elements in D, and this identity is an element of D) is regarded as part of the operational structure in the same way as having a multiplication on D or having inverses in D.